js_modules/dagster-ui/packages/ui-core/src/asset-data/AssetLiveDataProvider.tsx

```diff
@@ -0,0 +1,292 @@
+      id: `assetNodeLiveFragment-${assetNode.id}`,
+      data: assetNode,
+    });
+  });
+  data.assetsLatestInfo.forEach((assetLatestInfo) => {
+    _assetKeyToUniqueId[JSON.stringify(assetLatestInfo.assetKey.path)] = assetLatestInfo.id;
+    client.writeFragment({
+      fragment: ASSET_LATEST_INFO_FRAGMENT,
+      fragmentName: 'AssetLatestInfoFragment',
+      id: `assetLatestInfoFragment-${assetLatestInfo.id}`,
+      data: assetLatestInfo,
+    });
+  });
+}
+
+// How many assets to fetch at once
+const BATCH_SIZE = 50;
+
+// Milliseconds we wait until sending a batched query
+const BATCHING_INTERVAL = 100;
+
+export const SUBSCRIPTION_IDLE_POLL_RATE = 30 * 1000;
+// const SUBSCRIPTION_MAX_POLL_RATE = 2 * 1000;
+
+type DataForNodeListener = (stringKey: string, data: LiveDataForNode) => void;
+
+const AssetLiveDataContext = React.createContext<{
+  useAssetNodeLiveData: (keys: AssetKeyInput[]) => Record<string, LiveDataForNode | null>;
+}>({
+  useAssetNodeLiveData: () => ({}),
+});
+
+// Map of asset keys to their last fetched time and last requested time
+const lastFetchedOrRequested: Record<
+  string,
+  {fetched: number; requested?: undefined} | {requested: number; fetched?: undefined} | null
+> = {};
+
+export const AssetLiveDataProvider = ({children}: {children: React.ReactNode}) => {
+  const [needsImmediateFetch, setNeedsImmediateFetch] = React.useState<boolean>(false);
+
+  const client = useApolloClient();
+
+  function useAssetNodeLiveData(assetKeys: AssetKeyInput[]) {
+    const [data, setData] = React.useState<Record<string, LiveDataForNode | null>>({});
+
+    React.useEffect(() => {
+      const setDataSingle = (stringKey: string, assetData: LiveDataForNode) => {
+        setData((data) => {
+          return {...data, [stringKey]: assetData};
+        });
+      };
+      assetKeys.forEach((key) => {
+        _subscribeToAssetKey(client, key, setDataSingle, setNeedsImmediateFetch);
+      });
+      return () => {
+        assetKeys.forEach((key) => {
+          _unsubscribeToAssetKey(key, setDataSingle);
+        });
+      };
+    }, [assetKeys]);
+
+    return data;
+  }
+
+  const isDocumentVisible = useDocumentVisibility();
+
+  React.useEffect(() => {
+    if (!isDocumentVisible) {
+      return;
+    }
+    const interval = setInterval(() => {
+      _batchedQueryAssets(_determineAssetsToFetch(), client, (data) => {
+        Object.entries(data).forEach(([key, assetData]) => {
+          const listeners = _assetKeyListeners[key];
+          if (!listeners) {
+            return;
+          }
+          listeners.forEach((listener) => {
+            listener(key, assetData);
+          });
+        });
+      });
+      // Check for assets to fetch every 5 seconds to simplify logic
+      // This means assets will be fetched at most 5 + SUBSCRIPTION_IDLE_POLL_RATE after their first fetch
+      // but then will be fetched every SUBSCRIPTION_IDLE_POLL_RATE after that
+    }, 5000);
```

Comment:
It seems like with this change, we'll make a request every 5 seconds even if the requests take longer than 5s, which seems potentially bad -- the old implementation of this used `fetching.current` to make sure that we would never "stack" queries.

Would be good to test the new behavior by throwing a `time.sleep` in the live data resolver and making sure that this does not start stacking requests

Reply:
Every 5 seconds we'll make a request if `_determineAssetsToFetch()` determines that there are assets to fetch. In practice most of the time it will says there are no assets to fetch. I wrote a test case to cover this but I will be adding more to cover more cases